tutorials/haskell/snap-file-uploads/tutorial.md

```diff
@@ -0,0 +1,313 @@
+actually do something with the files. Here, we choose to return a list
+of either a `Text` (an error) or a `ByteString` (the contents of an
+uploaded file):
+
+```haskell
+handleFilesUpload' :: Snap [Either Text ByteString]
+handleFilesUpload' = do
+  temporaryDirectory <- MonadIO.liftIO Directory.getTemporaryDirectory
+  SnapFileUploads.handleFileUploads
+    temporaryDirectory
+    uploadPolicy
+    partUploadPolicy
+    handleFileRead
+```
+
+As temporary directory, we'll simply use the system's temporary
+directory.
+
+The general upload policy allows us to specify things like the maximum
+number of form inputs and the minimum upload rate. Snap provides a
+default upload policy, which is good enough for our application:
+
+```haskell
+uploadPolicy :: UploadPolicy
+uploadPolicy =
+  SnapFileUploads.defaultUploadPolicy
+```
+
+With the part or per-file upload policy, we allow or disallow each
+submitted file. Given a
+[`PartInfo`](https://hackage.haskell.org/package/snap-core-1.0.0.0/docs/Snap-Util-FileUploads.html#t:PartInfo),
+we either disallow a file to be uploaded or allow it with a maximum
+size. A value of `PartInfo` gives us information such as the content
+type of the file and the name of the file (which is optional and will be
+`Nothing` if we submit no files):
+
+```haskell
+partUploadPolicy
+  :: PartInfo
+  -> PartUploadPolicy
+partUploadPolicy partInfo =
+  if (SnapFileUploads.partContentType partInfo == "text/x-haskell"
+       || SnapFileUploads.partContentType partInfo == "text/x-literate-haskell")
+       && Maybe.isJust (SnapFileUploads.partFileName partInfo)
+     then
+       SnapFileUploads.allowWithMaximumSize maximumSize
+     else
+       SnapFileUploads.disallow
+```
+
+We only allow Haskell files (`.hs` or `.lhs`) with a name (no empty
+submissions) with a maximum size of one megabyte:
+
+```haskell
+maximumSize :: Int64
+maximumSize =
+  1 * megabyte
+  where
+    megabyte =
+      2 ^ (20 :: Int)
+```
+
+Note that this part upload policy is not very precise. We could rename a
+`turbulent-sniffle.pdf` file to `turbulent-sniffle.hs` and upload it.
+Also, the content type of a Haskell file could be submitted as
+`application/octet-stream`, which our upload policy would reject.
+
+At last, we can implement the `handleFileRead` and actually do something
+with the uploaded files. This handler takes a `PartInfo` and a file path
+(or a policy violation) and returns either an error or the contents of
+an uploaded file. We already used the `PartInfo` to allow or disallow
+the file, so we ignore it here:
+
+```haskell
+handleFileRead
+  :: PartInfo
+  -> Either PolicyViolationException FilePath
+  -> IO (Either Text ByteString)
+handleFileRead _ eitherFile =
+  case eitherFile of
+    Left policyViolation ->
+      return (Left (SnapFileUploads.policyViolationExceptionReason policyViolation))
+    Right file ->
+      fmap Right (ByteStringChar8.readFile file)
+```
+
+> Witchy kitchy kitchy wee,  
```

Comment:
@mrkkrp True. I didn't find anything good for this tutorial and the three unrelated snippets are indeed distracting. I removed them and also updated the paragraph about exporting the `main` function. Thanks!